Amplify/Categories/DataStore/Model/Temporal/Temporal+Cache.swift

```diff
@@ -0,0 +1,172 @@
+            formatter.timeZone = timeZone
+            return formatter
+            // defer takes care of unlock
+        }
+
+        // We're about to create a new formatter,
+        // so let's make sure we're being responsible
+        // about the amount of formatters we're caching.
+        // If we already have at least 15 formatters in the cache,
+        // we're going to evict one using a
+        // random replacement (RR) eviction policy
+        if formatterCache.count >= 15 {
+            // From the `randomElement()` documentation.
+            // "A random element from the collection. If the collection is empty, the method returns nil."
+            // Since we just confirmed the cache isn't empty, and we're guaranteeing exclusive access
+            // through the lock, we can safely force unwrap here.
+            let keyToEvict = formatterCache.randomElement()!.key
+            formatterCache[keyToEvict] = nil
+            // This can likely be improved at a later time by
+            // using a LFU, or potentially LRU, eviction policy.
+        }
+
+        // unlock if no early return
+        os_unfair_lock_unlock(&lock)
+
+        // Finally, if the formatter is not in the cache
+        // or a formatter with the matching format is cached,
+        // but the time zone doesn't match *and* the formatter
+        // is already referenced elsewhere, we create one.
+        let formatter = DateFormatter()
+        formatter.dateFormat = format
+        formatter.calendar = iso8601Calendar
+        formatter.locale = Locale(identifier: "en_US_POSIX")
+        formatter.timeZone = timeZone
+
+        // lock before write to cache
+        os_unfair_lock_lock(&lock)
+        formatterCache[format] = formatter
+        return formatter
+        // defer takes care of unlock
+    }
+
+    @usableFromInline
+    /// Turn a `String` into a `Foundation.Date`
+    /// - Parameters:
+    ///   - string: The date in `String` form.
+    ///   - formats: Any formats in `String` form that you want to check as a variadic argument.
+    ///   - timeZone: The `TimeZone` used by the `DateFormatter` when converted.
+    ///               Default is `.utc` a.k.a. `TimeZone(abbreviation: "UTC")`
+    /// - Returns: A `Foundation.Date` if conversion was successful.
+    /// - Throws: `DataStoreError.invalidDateFormat(_:)` if conversion was unsuccessful.
+    /// - SeeAlso: `date(from:with:in:)` overload that takes `[String]` rather than `String...`
+    internal static func date(
+        from string: String,
+        with formats: String...,
+        in timeZone: TimeZone = .utc
+    ) throws -> Foundation.Date {
+        for format in formats {
+            let formatter = formatter(for: format, in: timeZone)
+            if let date = formatter.date(from: string) {
+                return date
+            }
+        }
+        throw DataStoreError
+            .invalidDateFormat(formats.joined(separator: " | "))
```

Comment:
The variadic arg overload should actually just be removed at this point. It was added as a public convenience option. But since we've made these internal, there's no need for it anymore.